Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/PowerSupplies/PassengerCarPowerSupply.cs

```diff
@@ -0,0 +1,452 @@
+                    break;
+            }
+        }
+
+        public void Copy(IPowerSupply other)
+        {
+            if (other is ScriptedPassengerCarPowerSupply scriptedOther)
+            {
+                Copy(scriptedOther);
+            }
+        }
+
+        public void Copy(ScriptedPassengerCarPowerSupply other)
+        {
+            BatterySwitch.Copy(other.BatterySwitch);
+
+            ScriptName = other.ScriptName;
+
+            PowerOnDelayS = other.PowerOnDelayS;
+            ContinuousPowerW = other.ContinuousPowerW;
+            HeatingPowerW = other.HeatingPowerW;
+            AirConditioningPowerW = other.AirConditioningPowerW;
+            AirConditioningYield = other.AirConditioningYield;
+        }
+
+        public virtual void Initialize()
+        {
+            if (!Activated)
+            {
+                if (ScriptName != null && ScriptName != "Default")
+                {
+                    var pathArray = new string[] { Path.Combine(Path.GetDirectoryName(Wagon.WagFilePath), "Script") };
+                    Script = Simulator.ScriptManager.Load(pathArray, ScriptName) as PassengerCarPowerSupply;
+                }
+                if (Script == null)
+                {
+                    Script = new DefaultPassengerCarPowerSupply();
+                }
+
+                AssignScriptFunctions();
+
+                Script.Initialize();
+                Activated = true;
+            }
+
+            BatterySwitch.Initialize();
+        }
+
+        /// <summary>
+        /// Initialization when simulation starts with moving train
+        /// <\summary>
+        public virtual void InitializeMoving()
+        {
+            BatterySwitch.InitializeMoving();
+
+            ElectricTrainSupplyState = PowerSupplyState.PowerOn;
+            BatteryState = PowerSupplyState.PowerOn;
+
+            Script?.InitializeMoving();
+        }
+
+        public virtual void Save(BinaryWriter outf)
+        {
+            BatterySwitch.Save(outf);
+
+            outf.Write(FrontElectricTrainSupplyCableConnected);
+
+            outf.Write(ElectricTrainSupplyState.ToString());
+            outf.Write(LowVoltagePowerSupplyState.ToString());
+            outf.Write(BatteryState.ToString());
+        }
+
+        public virtual void Restore(BinaryReader inf)
+        {
+            BatterySwitch.Restore(inf);
+
+            FrontElectricTrainSupplyCableConnected = inf.ReadBoolean();
+
+            ElectricTrainSupplyState = (PowerSupplyState)Enum.Parse(typeof(PowerSupplyState), inf.ReadString());
+            BatteryState = (PowerSupplyState)Enum.Parse(typeof(PowerSupplyState), inf.ReadString());
+
+            firstUpdate = false;
+        }
+
+        public virtual void Update(float elapsedClockSeconds)
+        {
+            CarId = Train?.Cars.IndexOf(Wagon) ?? 0;
+
+            if (firstUpdate)
+            {
+                firstUpdate = false;
```

Comment:
Testing for the first update like this is brittle and unlikely to survive a simulator save/restore. Could we possibly move this logic to Initialize()?

Reply:
We can't:
The Train member is not initialized when the Initialize is called.